mmaction/core/evaluation/average_precision.py

```diff
@@ -0,0 +1,315 @@
+            classification_error, time_error, other_error and
+            missing_detections
+    """
+    ap_results = []
+    other_ap_results = [[], [], [], []]
+    missing_detections = []
+    for label_index, label in enumerate(labels):
+        det_result = det_results[det_results[:, 2] == label_index, :]
+
+        gt = defaultdict(list)
+        other_gt = defaultdict(list)
+        label_dict = defaultdict(list)
+
+        for video_id, video in enumerate(videos):
+            tubes = gt_tubes[video]
+            label_dict[video_id] = list(tubes)
+
+            for tube_label_index in tubes:
+                for tube in tubes[tube_label_index]:
+                    for i in range(tube.shape[0]):
+                        key = (video_id, int(tube[i, 0]))
+                        if tube_label_index == label_index:
+                            gt[key].append(tube[i, 1:5].tolist())
+                        else:
+                            other_gt[key].append(tube[i, 1:5].tolist())
+
+        for key in gt:
+            gt[key] = np.array(gt[key].copy())
+        for key in other_gt:
+            other_gt[key] = np.array(other_gt[key].copy())
+
+        original_key = list(gt)
+
+        precision_recall = np.empty((det_result.shape[0] + 1, 6),
+                                    dtype=np.float32)
+        precision_recall[0, 0] = 1.0
+        precision_recall[0, 1:] = 0.0
+
+        fn = sum([item.shape[0] for item in gt.values()])
+        (fp, tp, localization_error, classification_error, other_error,
+         time_error) = (0, 0, 0, 0, 0, 0)
+
+        for i, j in enumerate(np.argsort(-det_result[:, 3])):
+            key = (int(det_result[j, 0]), int(det_result[j, 1]))
+            box = det_result[j, 4:8]
+            is_positive = False
+
+            if key in original_key:
+                if key in gt:
+                    ious = iou2d(gt[key], box)
+                    max_idx = np.argmax(ious)
+
+                    if ious[max_idx] >= threshold:
+                        is_positive = True
+                        gt[key] = np.delete(gt[key], max_idx, 0)
+
+                        if gt[key].size == 0:
+                            del gt[key]
+                    else:
+                        localization_error += 1
+                else:
+                    localization_error += 1
+
+            elif key in other_gt:
+                ious = iou2d(other_gt[key], box)
+                if np.max(ious) >= threshold:
+                    classification_error += 1
+                else:
+                    other_error += 1
+
+            elif label_index in label_dict[key[0]]:
+                time_error += 1
+            else:
+                other_error += 1
+
+            if is_positive:
+                tp += 1
+                fn -= 1
+            else:
+                fp += 1
+
+            precision_recall[i + 1, 0] = tp / (tp + fp)
+            precision_recall[i + 1, 1] = tp / (tp + fn)
```

Comment:
any corner case making this divide by 0?

Reply:
1. `(tp + fp)` will not lead to 0, since the original `(tp +fp) == 0`. and then, either `tp+=1` or `fp+=1` , so `(tp+fp)` will be always positive.
2. `(tp + fn)` will not lead to 0, since the original `(tp +fn) > 0`. and then, either `tp+=1` and `fn-=1` happen at the same time , so `(tp+fn)` will be always positive.